patches/tModLoader/Terraria/Player.cs.patch

```diff
@@ -7787,56 +7787,66 @@
  			throw;
  		}
  	}
-@@ -41979,21 +_,52 @@
+@@ -41979,21 +_,60 @@
  		if (string.IsNullOrEmpty(path))
  			return;
  
-+		// TML: Delay vanilla saving until modded data is serialized (in case of exceptions)
-+		// Moved to the bottom
++		// TML: Split method into several methods to facilitate separation between saving to disk and generating save data.
++		// TML: Delay vanilla saving until modded data is serialized(in case of exceptions)
++		// Moved to UpdateLocalPlayerFile
 +		/*
  		if (FileUtilities.Exists(path, isCloudSave))
  			FileUtilities.Copy(path, path + ".bak", isCloudSave);
 +		*/
++		PlayerLoader.PreSavePlayer(player);
  
- 		RijndaelManaged rijndaelManaged = new RijndaelManaged();
++		byte[] plrData = GeneratePlayerData(playerFile);
 +
-+		/*
- 		using Stream stream = (isCloudSave ? ((Stream)new MemoryStream(2000)) : ((Stream)new FileStream(path, FileMode.Create)));
-+		*/
++		PlayerLoader.PostSavePlayer(player);
++
++		TagCompound tplrData = PlayerIO.SaveData(player);
++
++		UpdateLocalPlayerFile(playerFile, plrData, tplrData);
++	}
++
++	public static byte[] GeneratePlayerData(PlayerFileData playerFile)
++	{
+ 		RijndaelManaged rijndaelManaged = new RijndaelManaged();
+-		using Stream stream = (isCloudSave ? ((Stream)new MemoryStream(2000)) : ((Stream)new FileStream(path, FileMode.Create)));
 +		using Stream stream = new MemoryStream(2000);
  		using CryptoStream cryptoStream = new CryptoStream(stream, rijndaelManaged.CreateEncryptor(ENCRYPTION_KEY, ENCRYPTION_KEY), CryptoStreamMode.Write);
  		using BinaryWriter binaryWriter = new BinaryWriter(cryptoStream);
-+
-+		PlayerLoader.PreSavePlayer(player);
-+
  		binaryWriter.Write(279);
  		playerFile.Metadata.Write(binaryWriter);
- 		Serialize(playerFile, player, binaryWriter);
+-		Serialize(playerFile, player, binaryWriter);
++		Serialize(playerFile, playerFile.Player, binaryWriter);
```

Comment:
Better to copy `var player = playerFile.Player` into the top of the method, to keep all the patches together with the method split/change